### PR TITLE
[TR] PIM-4059: Rework categories

### PR DESCRIPTION
**Pull request by grena**

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/ProductCategoryRepository.php [hunk 1]
```diff
@@ -236,4 +187,79 @@ public function applyFilterByCategoryIdsOrUnclassified($qb, array $categoryIds)
                 ->addOr($qb->expr()->field('categoryIds')->size(0))
         );
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItemCountByTree($product)
+    {
+        $categories = $product->getCategories();
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/ProductCategoryRepository.php]: if not instanceof ProductInterface 
throw invalid arg exception ma gueule

and add the exception in the interface please

File: src/Pim/Bundle/FilterBundle/Filter/CategoryFilter.php [hunk 2]
```diff
@@ -1,18 +1,16 @@
 <?php
 
-namespace Pim\Bundle\FilterBundle\Filter\Product;
+namespace Pim\Bundle\FilterBundle\Filter;
```
Comment (jjanvier) [inline src/Pim/Bundle/FilterBundle/Filter/CategoryFilter.php]: missing in changelog

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/ProductCategoryRepository.php [hunk 3]
```diff
@@ -236,4 +191,88 @@ public function applyFilterByCategoryIdsOrUnclassified($qb, array $categoryIds)
...
+            throw new \InvalidArgumentException(
+                sprintf(
+                    'Expected a "Pim\Bundle\CatalogBundle\Model\ProductInterface", got a "%s"',
+                    get_class($product)
+                )
+            );
+        }
+
+        $categories = $product->getCategories();
+        $categoryIds = array();
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/ProductCategoryRepository.php]: use []

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/ProductCategoryRepository.php [hunk 4]
```diff
@@ -236,4 +191,88 @@ public function applyFilterByCategoryIdsOrUnclassified($qb, array $categoryIds)
...
+                   "    tree.id AS tree_id,".
+                   "    '0' AS product_count".
+                   "  FROM $categoryTable tree".
+                   "  LEFT JOIN $categoryTable category".
+                   "    ON category.root = tree.id".
+                   " WHERE tree.parent_id IS NULL".
+                   " GROUP BY tree.id";
+        }
+
+        $stmt = $this->entityManager->getConnection()->prepare($sql);
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/ProductCategoryRepository.php]: Why do not use DQL ?

File: src/Pim/Bundle/CatalogBundle/Manager/CategoryManager.php [hunk 5]
```diff
@@ -21,22 +26,38 @@ class CategoryManager
...
+     * @param CategoryRepositoryInterface $productCategoryRepo
+     * @param CategoryFactory             $categoryFactory
      */
-    public function __construct(ObjectManager $om, $categoryClass)
-    {
-        $this->om = $om;
-        $this->categoryClass = $categoryClass;
+    public function __construct(
+        ObjectManager $om,
+        $categoryClass,
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Manager/CategoryManager.php]: as you change the constructor, maybe it's ok to put this argument last

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductCategoryRepository.php [hunk 6]
```diff
@@ -195,4 +136,98 @@ public function applyFilterByCategoryIdsOrUnclassified($qb, array $categoryIds)
...
+        $stmt = $this->em->getConnection()->prepare($sql);
+        $stmt->bindValue('productId', $product->getId());
+
+        $stmt->execute();
+        $productCounts = $stmt->fetchAll();
+        $trees = array();
+        foreach ($productCounts as $productCount) {
+            $tree = array();
+            $tree['productCount'] = $productCount['product_count'];
+            $tree['tree'] = $this->em->getRepository($categoryClass)->find($productCount['tree_id']);
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductCategoryRepository.php]: Can you create a variable `$categoryRepo = $this->em->getRepository($categoryClass);` before this `foreach` please ?

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductCategoryRepository.php [hunk 7]
```diff
@@ -195,4 +136,98 @@ public function applyFilterByCategoryIdsOrUnclassified($qb, array $categoryIds)
...
+            "    ON category_product.product_id = :productId".
+            "   AND category_product.category_id = category.id".
+            " GROUP BY tree.id";
+
+        $stmt = $this->em->getConnection()->prepare($sql);
+        $stmt->bindValue('productId', $product->getId());
+
+        $stmt->execute();
+        $productCounts = $stmt->fetchAll();
+        $trees = array();
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductCategoryRepository.php]: use []

File: src/Pim/Bundle/CatalogBundle/Manager/CategoryManager.php [hunk 8]
```diff
@@ -197,12 +231,12 @@ public function getTreeByCode($code)
...
      * @return CategoryInterface
+     *
+     * @deprecated Please use CategoryRepositoryInterface::findOneBy() instead
      */
     public function getCategoryByCode($code)
     {
-        return $this
-            ->getEntityRepository()
-            ->findOneBy(array('code' => $code));
+        return $this->productCategoryRepo->findOneBy(array('code' => $code));
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Manager/CategoryManager.php]: use []

File: src/Pim/Bundle/CatalogBundle/Repository/ProductCategoryRepositoryInterface.php [hunk 9]
```diff
@@ -21,6 +21,8 @@
      * @param ProductInterface $product The product to look for in the trees
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Repository/ProductCategoryRepositoryInterface.php]: ProductCategoryRepositoryInterface should extend ItemCategoryInterface and FilterableCategoryInterface in 1.5, please add a TODO

File: src/Pim/Component/Classification/Repository/CategoryFilterableRepositoryInterface.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+namespace Pim\Component\Classification\Repository;
+
+/**
+ * @author    Adrien Pétremann <adrien.petremann@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface CategoryFilterableRepositoryInterface
+{
+    /**
```
Comment (jjanvier) [inline src/Pim/Component/Classification/Repository/CategoryFilterableRepositoryInterface.php]: no applyFilterByIds here ?

File: CHANGELOG.md [hunk 11]
```diff
@@ -144,6 +144,13 @@
...
+- Change `Pim\Bundle\CatalogBundle\Manager\CategoryManager`, updated method `getEntityRepository` to return return a `Pim\Component\Classification\Repository\CategoryRepositoryInterface`
+- Change constructor of `Pim\Bundle\EnrichBundle\Controller\CategoryTreeController`, added `Pim\Bundle\CatalogBundle\Factory\CategoryFactory` and `Pim\Component\Classification\Repository\CategoryRepositoryInterface` as last arguments
+- Change constructor of `Pim\Bundle\EnrichBundle\Controller\ProductController`, added `Pim\Bundle\CatalogBundle\Factory\CategoryFactory` as last argument 
+- Move `Pim\Bundle\FilterBundle\Filter\Product\CategoryFilter` to `Pim\Bundle\FilterBundle\Filter\CategoryFilter`
+- Change constructor of `Pim\Bundle\FilterBundle\Filter\CategoryFilter`, last argument is now a `Pim\Component\Classification\Repository\CategoryRepositoryInterface`
+- Change constructor of `Pim\Bundle\CatalogBundle\Doctrine\Common\Filter\CategoryFilter`, second argument is now a `Pim\Component\Classification\Repository\CategoryFilterableRepositoryInterface`
```
Comment (jjanvier) [inline CHANGELOG.md]: missing BC breaks: some functions have been added to ProductCategoryRepo

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/tree-view.jstree.js [hunk 12]
```diff
@@ -130,7 +131,7 @@ define(
                             return {
                                 id: getNodeId(node),
                                 select_node_id: getActiveNode(),
-                                with_products_count: 1,
+                                with_products_count: 1, // TODO: Change naming
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/tree-view.jstree.js]: you don't do it in this pr ?

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php [hunk 13]
```diff
@@ -12,24 +13,27 @@
  * @author    Romain Monceau <romain@akeneo.com>
  * @copyright 2013 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *
+ * TODO: A lot of protected functions can be moved into a proper repository
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: if you don't do it here, add this in a task of a story please

File: src/Pim/Bundle/UIBundle/Resources/translations/jsmessages.en.yml [hunk 14]
```diff
@@ -43,7 +43,11 @@ Unit:     Unit
 Currency: Currency
 
 jstree:
-    all:          All products
-    unclassified: Unclassified products
+    product:
+        all:          All products
+        unclassified: Unclassified products
+    asset:
```
Comment (jjanvier) [inline src/Pim/Bundle/UIBundle/Resources/translations/jsmessages.en.yml]: nope, it's for EE :p

File: src/Pim/Component/Classification/Repository/CategoryFilterableRepositoryInterface.php (hunk 10 above)
Comment (jjanvier) [inline src/Pim/Component/Classification/Repository/CategoryFilterableRepositoryInterface.php]: no filterByIds here ?

File: src/Pim/Component/Classification/Updater/CategoryUpdater.php [hunk 15]
```diff
@@ -61,7 +62,7 @@ public function update($category, array $data, array $options = [])
      */
     protected function setData(CategoryInterface $category, $field, $data)
     {
-        if ('labels' === $field) {
+        if ('labels' === $field && $category instanceof TranslatableInterface) {
```
Comment (jjanvier) [inline src/Pim/Component/Classification/Updater/CategoryUpdater.php]: arf :(

but OK, no other choice atm

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/ProductCategoryRepository.php (hunk 4 above)
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/ProductCategoryRepository.php]: it's just a copy pasta, nothing new here

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductCategoryRepository.php (hunk 6 above)
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductCategoryRepository.php]: Copy pasta again, but yes I'll do it while I'm here :)

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/ProductCategoryRepository.php [hunk 16]
```diff
@@ -236,4 +191,88 @@ public function applyFilterByCategoryIdsOrUnclassified($qb, array $categoryIds)
...
+    /**
+     * {@inheritdoc}
+     */
+    public function getItemCountByTree($product)
+    {
+        if (!$product instanceof ProductInterface) {
+            throw new \InvalidArgumentException(
+                sprintf(
+                    'Expected a "Pim\Bundle\CatalogBundle\Model\ProductInterface", got a "%s"',
+                    get_class($product)
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/ProductCategoryRepository.php]: Please use Doctrine\Common\Util\ClassUtils::getClass()

File: src/Pim/Component/Classification/Factory/CategoryFactory.php [hunk 17]
```diff
@@ -1,13 +1,11 @@
 <?php
```
Comment (phaseinducer) [inline src/Pim/Component/Classification/Factory/CategoryFactory.php]: No interface for this factory ?

File: src/Pim/Bundle/CatalogBundle/Manager/CategoryManager.php (hunk 5 above)
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Manager/CategoryManager.php]: yes, good idea

File: src/Pim/Bundle/CatalogBundle/Manager/CategoryManager.php [hunk 18]
```diff
@@ -21,22 +26,38 @@ class CategoryManager
...
+        $this->productCategoryRepo = $productCategoryRepo;
+        $this->categoryFactory     = $categoryFactory;
     }
 
     /**
      * Return object manager
      *
      * @return ObjectManager
+     *
+     * @deprecated
```
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Manager/CategoryManager.php]: please add "will be removed in 1.5"

File: src/Pim/Bundle/CatalogBundle/Manager/CategoryManager.php [hunk 19]
```diff
@@ -12,6 +15,8 @@
  * @author    Romain Monceau <romain@akeneo.com>
  * @copyright 2013 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *
+ * @deprecated Will be removed in 1.5
```
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Manager/CategoryManager.php]: and what should we use? ;)

File: src/Pim/Bundle/CatalogBundle/Resources/config/factories.yml [hunk 20]
```diff
@@ -47,7 +46,7 @@ services:
             - %pim_catalog.entity.group.class%
 
     pim_catalog.factory.category:
-        class: %pim_catalog.factory.category.class%
+        class: %pim_classification.factories.category.class%
```
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Resources/config/factories.yml]: rename the parameter to pim_classification.factory.category.class

File: src/Pim/Bundle/EnrichBundle/Controller/CategoryTreeController.php [hunk 21]
```diff
@@ -96,11 +110,13 @@ public function __construct(
             $doctrine
         );
 
-        $this->categoryManager = $categoryManager;
-        $this->userContext     = $userContext;
-        $this->securityFacade  = $securityFacade;
-        $this->categorySaver   = $categorySaver;
-        $this->categoryRemover = $categoryRemover;
+        $this->categoryManager    = $categoryManager;
```
Comment (nidup) [inline src/Pim/Bundle/EnrichBundle/Controller/CategoryTreeController.php]: not far to be able to not use it at all and to depreciate it, greeeeaaat job!

File: src/Pim/Bundle/UIBundle/Resources/translations/jsmessages.en.yml (hunk 14 above)
Comment (nidup) [inline src/Pim/Bundle/UIBundle/Resources/translations/jsmessages.en.yml]: never drop translation keys! we have to keep the old one, if not we'll loose them in older versions

File: src/Pim/Component/Classification/Updater/CategoryUpdater.php (hunk 15 above)
Comment (nidup) [inline src/Pim/Component/Classification/Updater/CategoryUpdater.php]: Asset Categories will not have labels? if yes, put a TODO

File: src/Pim/Bundle/FilterBundle/Filter/CategoryFilter.php (hunk 2 above)
Comment (grena) [inline src/Pim/Bundle/FilterBundle/Filter/CategoryFilter.php]: Oups, in fact it is:
https://github.com/akeneo/pim-community-dev/pull/2862/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR151

File: src/Pim/Bundle/CatalogBundle/Repository/ProductCategoryRepositoryInterface.php (hunk 9 above)
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Repository/ProductCategoryRepositoryInterface.php]: done

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php (hunk 13 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: done

File: src/Pim/Component/Classification/Updater/CategoryUpdater.php (hunk 15 above)
Comment (grena) [inline src/Pim/Component/Classification/Updater/CategoryUpdater.php]: They will, but the `CategoryInterface` from ClassificationBundle is not aware of locales, the `setLocale` method comes with the `TranslatableInterface`.
 Isn't it a good point?

Comment (jjanvier) [inline src/Pim/Component/Classification/Updater/CategoryUpdater.php]: the bad point is that TranslatableInterface comes from a bundle

File: src/Pim/Component/Classification/Factory/CategoryFactory.php (hunk 17 above)
Comment (grena) [inline src/Pim/Component/Classification/Factory/CategoryFactory.php]: Nop indeed, I don't know if it's really useful for a factory?

**Pull request merged**